_posts/2025-05-05-plan-execute-agent.md

```diff
@@ -0,0 +1,423 @@
+
+## Extending the Agent with More Tools and MCP Support
+
+The power of the Plan–Execute–And-Reflect agent lies in its flexibility. While our example uses a basic set of tools to investigate a cart failure, the agent can support a wide range of tools. For the list of available tools, see [Tools](https://docs.opensearch.org/docs/latest/ml-commons-plugin/agents-tools/tools/index/).
+
+One such example would be to use the [WebSearchTool](https://docs.opensearch.org/docs/latest/ml-commons-plugin/agents-tools/tools/web-search-tool/) to perform deep-research tasks. 
+
+In addition, this agent acts as an MCP client, meaning it can connect to an MCP (Model Coordination Protocol) server. This allows the agent to retrieve tools and configurations dynamically and participate in more complex workflows.
+
+If you’d like to connect your agent to an MCP server, refer to this [document](https://docs.opensearch.org/docs/latest/ml-commons-plugin/agents-tools/mcp/mcp-connector/) for detailed instructions.
+
+### Recommendations and troubleshooting
+
+* If you notice your agent getting throttled by your model, enable retries via the [connector configuration](https://docs.opensearch.org/docs/latest/ml-commons-plugin/remote-models/blueprints/#configuration-parameters).
+* If the agent stops with max_steps executed, then you can increase the [max_steps](https://docs.opensearch.org/docs/latest/ml-commons-plugin/api/agent-apis/register-agent/#request-body-fields) or specify it during execution via the `parameters.max_steps` field.
+
+Note: This is an experimental feature and will continue to evolve with future enhancements.
+
+### Future Enhancements
+
+* Support parallel tool execution
+* Support multiple reflection strategies
+* Human in the loop
+* Ability to cancel a running task
+* Agent execution checkpointing 
+
+For updates on the progress of the feature or if you want to leave feedback, see the associated GitHub [issue](https://github.com/opensearch-project/ml-commons/issues/3745).
+
+## Conclusion
+
+In this tutorial, we demonstrated how to use the Plan–Execute–Reflect agent to troubleshoot and identify the root cause of a cart failure in the opentelemetry-demo application. By breaking down a high-level task into smaller steps and executing them iteratively, the agent provides a powerful and flexible way to automate complex reasoning workflows. This approach is particularly useful for observability and operational intelligence use cases where multi-step diagnosis is needed.
+
+The Plan–Execute–Reflect agent demonstrates how intelligent automation can simplify root cause analysis. By combining OpenTelemetry observability data with LLM-based reasoning, OpenSearch can now assist engineers in debugging complex, multi-service systems with just a single prompt.
+
+Want to try it yourself? Check out this [tutorial](https://docs.opensearch.org/docs/latest/tutorials/gen-ai/agents/build-plan-execute-reflect-agent/) and deploy the agent in your own stack.
```

Comment:
I feel we should not mention the same tutorial again here, as it's mentioned earlier 
> For more details on agent registration and execution, see this [tutorial](https://docs.opensearch.org/docs/latest/tutorials/gen-ai/agents/build-plan-execute-reflect-agent/).

Any suggestion from tech doc writer ?

Reply:
@ylwu-amzn I think it's OK to mention it again here because we're including it for a different purpose (a call to action to the reader) and the two mentions are sufficiently far apart in the post, but I'll let you, @pyek-bot, and @kolchfa-aws make the call 😄

Reply:
Right. I would keep this as a call to action for a quick next step once the reader is finished with the blog.